### PR TITLE
Refactor Transitions and Moves

### PR DESCRIPTION
**Pull request by Adda0**
This PR refactors Moves and Transitions according to what we have agreed on. Namely, it:
- refactor transition and move iterators into a simple single class objects with their own const iterators. 
- includes epsilon and symbol move iterators.
- updates Python binding to work with the current implementation over Transitions.
- moves methods about iterating over transitions into Transitions.

Questions to resolve:
- Do we need `Delta::transitions_begin()` (`..._end()`) and `StatePost::moves_begin()` (`epsilon_moves_...`, `symbol_moves_...`; `..._end()`) methods? Currently, you have to get the `Moves` and `Transitions` classes to get begin or end iterator to moves and transitions.

### CODE REVIEW

**Comment by Adda0** [comment]
@vhavlena Since you have been invested in the discussion about transitions and moves, could you have a look at how the transitions and moves were refactored (what is the current interface for them) and give your opinion on how you are satisfied with it?

**Comment by vhavlena** [comment]
It seems fine. I would rename `const_iterator` to `ConstIterator` to match the convention.

File: src/nfa/delta.cc [hunk 1]
```diff
@@ -59,12 +59,9 @@
     return state_post.end();
 }
 
-size_t Delta::size() const {
-    size_t size = 0;
-    for (State q = 0; q < num_of_states(); ++q) {
-        for (const SymbolPost & m: (*this)[q]) { size = size + m.size(); }
-    }
-    return size;
+Delta& Delta::operator=(const Delta& other) {
```
Comment (kilohsakul) [inline src/nfa/delta.cc]: =default??

File: include/mata/nfa/delta.hh [hunk 2]
```diff
@@ -121,64 +121,105 @@ public:
...
-        moves_const_iterator begin() const { return begin_; }
-        moves_const_iterator end() const { return end_; }
-    };
+        /**
+         * @brief Whether to look up first symbol to iterate over from the beginning or from the end of @c StatePost
+         *  iterator.
+
+         * Iterates over symbol posts with symbols in interval [first_symbol_, last_symbol_].
+         */
+        enum class FirstSymbolLookupDiretion {
```
Comment (tfiedor) [inline include/mata/nfa/delta.hh]:
```suggestion
        enum class FirstSymbolLookupDirection {
```
Maybe `LookupDirection` would suffice? Or `IterationDirection`?

File: include/mata/nfa/delta.hh [hunk 3]
```diff
@@ -121,64 +121,105 @@ public:
...
+
+        /**
+         * Iterator over moves.
+         */
+        class const_iterator {
+        private:
+            const StatePost* state_post_{ nullptr };
+            StatePost::const_iterator state_post_it_{};
+            StateSet::const_iterator symbol_post_it_{};
+            Symbol last_symbol_{ Limits::max_symbol };
```
Comment (tfiedor) [inline include/mata/nfa/delta.hh]: I wonder if `sentinel_` would be appropriate name? It's short, elegant, well-established.

**Review by tfiedor — APPROVED** [review]
No Major nor Colonel Issues.

File: include/mata/nfa/delta.hh [hunk 4]
```diff
@@ -121,64 +121,105 @@
     const_iterator find(const Symbol symbol) const { return super::find({ symbol, {} }); }
 
     /**
-     * Iterator over moves. It iterates over pairs (symbol, target state) for a current source state whose state post
-     *  we iterate.
+     * @brief Iterator over moves represented as @c Move instances.
+     *
+     * It iterates over pairs (symbol, target) for the given @c StatePost.
      */
-    struct moves_const_iterator {
```
Comment (kilohsakul) [inline include/mata/nfa/delta.hh:127]: Btw, did we want to make state post with a member instead of inheritance from OrdVector?

Comment (Adda0) [reply include/mata/nfa/delta.hh:127]: Yes. We will refactor `StatePost` to have a member `symbol_posts` of type `OrdVector` instead of having it inherit from `Ordvector` for more explicit code which does not utilize more advanced features of C++ (inheritance of classes). It might be easier for students to read and understand the code with members instead of inheritance.

File: include/mata/nfa/delta.hh [hunk 5]
```diff
@@ -366,68 +409,75 @@
...
-    struct transitions_const_iterator {
-    private:
-        const std::vector<StatePost>& post_;
-        size_t current_state_;
-        StatePost::const_iterator post_iterator_{};
-        StateSet::const_iterator targets_position_{};
-        bool is_end_;
-        Transition transition_{};
-
+    class Transitions {
```
Comment (kilohsakul) [inline include/mata/nfa/delta.hh]: Hm, I am still wandering whether having all this code just to avoid one for loop when printing out automaton is worth it ... I know that it is stupid to say this after it is done.

File: src/nfa/delta.cc (hunk 1 above)
Comment (Adda0) [reply src/nfa/delta.cc]: It could not be default implemented when we had transitions as a member of Delta. When we no longer have that, it can be default implemented again using the member-wise copy assignment.

Comment (Adda0) [reply src/nfa/delta.cc]: Fixed.

File: include/mata/nfa/delta.hh [hunk 6]
```diff
@@ -121,64 +121,105 @@
...
+            Backward ///< Lookup from the end to the beginning.
+        };
+
+        /**
+         * Iterator over moves.
+         */
+        class const_iterator {
+        private:
+            const StatePost* state_post_{ nullptr };
+            StatePost::const_iterator state_post_it_{};
```
Comment (kilohsakul) [inline include/mata/nfa/delta.hh]: Names of state_post_it and symbol_post_it, may it be better to name them by to what they are dereferenced to instead of what class owns them? Maybe something as symbol_it and target_it? I am thinking about this since I find the code with the current names a little hard to read.

File: include/mata/nfa/delta.hh (hunk 3 above)
Comment (kilohsakul) [reply include/mata/nfa/delta.hh]: In act I would not know what sentinel means.

File: include/mata/nfa/delta.hh (hunk 2 above)
Comment (Adda0) [reply include/mata/nfa/delta.hh]: It is not an iteration direction. The moves are still iterated in the order of first to last, but the iteration starts at some first symbol which must be found in the vector of `SymbolPost`s first. The process of finding the iterator to this first symbol (and its `SymbolPost`) can proceed either backward (from the last element in the vector to the first one) or forward (from the first element to the last one). When you are looking for the first epsilon in the vector, you want to start from the end. Otherwise, you should start from the beginning of the vector.

File: include/mata/nfa/delta.hh (hunk 3 above)
Comment (tfiedor) [reply include/mata/nfa/delta.hh]: I believe, it is quite well-established naming for our `zarážka` in programming (e.g. in some lookup algorithm, I am fairly sure it is used). `last_symbol` sounds like something at `.back()`, which I'm not sure if applies for this particular case.

File: include/mata/nfa/delta.hh (hunk 2 above)
Comment (tfiedor) [reply include/mata/nfa/delta.hh]: OK.

File: include/mata/nfa/delta.hh (hunk 3 above)
Comment (Adda0) [reply include/mata/nfa/delta.hh]: To my best knowledge, it is as @tfiedor says. However, in this case I meant to truly use the notion of first symbol and last symbol to explicitly state that we iterate over a closed interval `[fist_symbol, last_symbol]`, that is, both first symbol and last symbol are iterated over. Sentinel is what we usually use for one after the last element, it is basically what we call the end iterator in C++ iterator semantics.

Comment (tfiedor) [reply include/mata/nfa/delta.hh]: How about just `max_symbol` then?

Comment (tfiedor) [reply include/mata/nfa/delta.hh]: My issue is, that it is little ambiguous (could be last as in position, or last as last seen, or last as upper limit, etc.)

Comment (Adda0) [reply include/mata/nfa/delta.hh]: `max_symbol` is fine with me. When talking about intervals, you use `[first, last]`, however, instead of `[first, max]`, no?

But I agree that in order to avoid ambiguous code, `max` might be easier to work with.

File: include/mata/nfa/delta.hh (hunk 6 above)
Comment (Adda0) [reply include/mata/nfa/delta.hh]: You either name the iterators after whose iterator it is (as in this case) or over what data the iterator iterates (in this case, it would be `symbol_post_it` iterating over symbol posts). I am fine with either, and I think both approaches are used equally. If you both think it would be better, I am all for changing the name as you suggest.

File: include/mata/nfa/delta.hh (hunk 5 above)
Comment (Adda0) [reply include/mata/nfa/delta.hh]: In that case, what about the class `Transition` and functions accepting this class? We would have no way to create an instance of `Transition`, so then the whole class would be pointless and could be removed, too, no?

File: include/mata/nfa/delta.hh (hunk 3 above)
Comment (jurajsic) [reply include/mata/nfa/delta.hh]: For me `max` is more ambigious than `last`, especially when it is initialized to `Limits::max_symbol`. I think a comment here would be good.

File: include/mata/nfa/delta.hh (hunk 5 above)
Comment (jurajsic) [reply include/mata/nfa/delta.hh]: I think having an iterator over all transitions can be useful in some other contexts than printing out automaton. For example if I want to check if automaton accepts only one symbol words, I can iterate all transitions and check if source state is initial and target is final (I feel like we did something like that in Noodler).

File: include/mata/nfa/delta.hh [hunk 7]
```diff
@@ -444,7 +494,9 @@ public:
      * @return An iterator to @c SymbolPost with epsilon symbol. End iterator when there are no epsilon transitions.
      */
     static StatePost::const_iterator epsilon_symbol_posts(const StatePost& state_post, Symbol epsilon = EPSILON);
-}; // struct Delta.
+private:
+    std::vector<StatePost> state_posts_;
+}; // class Delta.
```
Comment (jurajsic) [inline include/mata/nfa/delta.hh:445]: Why was this moved to the end of class definition?

**Review by jurajsic — APPROVED** [review]

File: include/mata/nfa/delta.hh (hunk 6 above)
Comment (Adda0) [reply include/mata/nfa/delta.hh]: It was decided that we will name iterators as `<the_name_of_the_object_we_get_by_dereferencing_the_iterator>_it`. We can slowly begin renaming all existing iterator variables to adhere to this convention.

For example, for const iterator over `StatePost`, we get the following variable name:
```cpp
StatePost::const_iterator symbol_post_it = nfa.delta.state_post(4).begin();
```

Comment (Adda0) [reply include/mata/nfa/delta.hh]: For further discussion, please refer to https://github.com/VeriFIT/mata/issues/313.

File: include/mata/nfa/delta.hh (hunk 5 above)
Comment (kilohsakul) [reply include/mata/nfa/delta.hh]: ok, lets keep it

File: include/mata/nfa/delta.hh (hunk 2 above)
Comment (kilohsakul) [reply include/mata/nfa/delta.hh]: We discussed and agreed to change it, this class will be gone, but I forgot what exactly should happen.

Comment (Adda0) [reply include/mata/nfa/delta.hh]: We will have only one function, `StatePost::moves()` which will take 3 optional parameters. Without any, it will behave as a normal moves iterator iterating over all moves (including epsilons) and when the parameters are specified, it will iterate over some subset of these: `StatePost::moves(moves_kind, first_symbol_iterator, last_symbol_iterator)` where `moves_kind` is of type enum class with values `MovesKind::All`, `MovesKind::Epsilons` and `MovesKind::Symbols`. Iterators then only specify, which subset of moves is to be iterated over. The kind specifies whether to look for the first symbol from the left or from the right. This replaces `FirstSymbolLookupDirection` class.

File: include/mata/nfa/delta.hh (hunk 3 above)
Comment (Adda0) [reply include/mata/nfa/delta.hh]: So what about keeping `last_symbol`, but adding an explanation, that it is by default the maximal symbol and that it means the last symbol in the vector, not the last + 1. Would that be sufficient?

File: include/mata/nfa/delta.hh (hunk 2 above)
Comment (tfiedor) [reply include/mata/nfa/delta.hh]: Sounds much better. Moreover, I though of few alternatives for `MovesKinds` (sounds kinda weird): `Pick::All`, `Select::Epsilons` or maybe `Iterate::Symbols`? (I wrote them with the second part as well, so one can see, how it sounds). Feel free to pick one, if you like any.

Comment (Adda0) [reply include/mata/nfa/delta.hh]: I am experimenting with the names and `Iterate::Symbols` seems good to me. I will use `Iterate` and others can give their opinion in the PR when the changes are pushed.

Comment (kilohsakul) [reply include/mata/nfa/delta.hh]: "Iterate" does not give much hint, no? 
But I don't know what it should be, IterateOver? SymbolKind / SymbolType....

Related to this, we should fix what "symbol" means. Currently we have the data type Symbol, that can be alphabet symbol or epsilon, but here David was using symbol as alphabet symbol only.
We could have symbol, alphabet symbol, epsilon, or perhaps better "Label" (for anything), "Symbol", and "Epsilon".

Comment (Adda0) [reply include/mata/nfa/delta.hh]: I think it explains it well: `nfa.delta[4].moves(Iterate::Epsilons)` gives me all the information I need, in my opinion. I personally do not see a difference between `Iterate` and `IterateOver`. But I can change it to `IterateOver` if you think that is better. Either sounds good to me.

Regarding the `Symbol` naming. You said that you do not consider epsilon to be a symbol, and want the “normal” symbols to be called just symbols. Therefore, the name gets a bit ambiguous. I think we can think about possible names, discuss it tomorrow in our meeting and later refactor the whole Mata accordingly. For now, I will use “alphabet symbols” for “normal” symbols and “epsilons” for epsilon symbols.

Comment (tfiedor) [reply include/mata/nfa/delta.hh]: > "Iterate" does not give much hint, no? But I don't know what it should be, IterateOver? SymbolKind / SymbolType....

I think, here, you have to take into the account the context and the usage. Since, this will be probably called as `*.moves(Iterate::Symbols, ...)`, just simple verb with the noun is IMO fine.

Comment (kilohsakul) [reply include/mata/nfa/delta.hh]: Hm, maybe one could use it, but is it sounds ilogical to me. If I read *.moves(bla) as a constructor of a set of moves 
(is this right? The set is not explicitly constructed, only iterable, but one still sees it as a set of moves.) 
then the bla should specify a property of the elements I want to include in the set. Iterate is not a porperty of an element, moves(Iterate:bla) would look like constructing the set of all moves with s special iteration that only considers its subset, weird.

Comment (tfiedor) [reply include/mata/nfa/delta.hh]: Whatever, I was thinking that it is iterator, and hence its parameters would specify properties of the iteration not its elements, but this discussion is getting nowhere, so go for the `MovesKind`.

Comment (Adda0) [reply include/mata/nfa/delta.hh]: >  (is this right? The set is not explicitly constructed, only iterable, but one still sees it as a set of moves.)

It is right.

> then the bla should specify a property of the elements I want to include in the set.

I do not agree. It can be just as well a property of the function (the constructor) which specifies more closely what to construct. However, I have no problems with `IteraveOver`. Are you OK with `IterateOver`? `state_post.moves(IterateOver::Epsilons)`?

Comment (Adda0) [reply include/mata/nfa/delta.hh]: We agreed now to keep the original three methods to create `Moves`. Therefore, we do not need this parameter to pick which symbol posts we want to iterate over. This parameter will be changed back to `SeekFirstSymbolDirection::Forward` and will be used by the user only when they want to explicitly fill all parameters themselves: `moves(first_symbol, last_symbol, direction)`. Otherwise, in 99 % of cases, they will just use one of the three functions without needing to specify any parameter.

Comment (kilohsakul) [reply include/mata/nfa/delta.hh]: Wait, what is the SeekFirstSymbolDirection::Forward ? That should be completely gone, no?

Comment (Adda0) [reply include/mata/nfa/delta.hh]: You said in our discussion in Discord to rename `Iterate` to `SeekFirst` or something similar to what I had earlier. I can rename it to something else, but the meaning stays. It no longer represents a kind of moves to create. We have named functions for that. It only means where to start looking for the first symbol. What would you prefer?

Alternatively, I can completely remove this optimization seeing that it brings us so much headache and just look for the first symbol using binary search. Only after checking whether the first and last symbol posts in the state post are already what we look for.

File: include/mata/nfa/delta.hh (hunk 3 above)
Comment (Adda0) [reply include/mata/nfa/delta.hh]: I added more explanations. Hopefully, the distinction will now be clear.

File: include/mata/nfa/delta.hh (hunk 7 above)
Comment (Adda0) [reply include/mata/nfa/delta.hh:445]: I needed more private members earlier (which are not needed now), which had to be defined only after their types have been defined in the Delta class. Therefore, the private members list had to be at the end of the class definition. In the end, only the original member remained, but private members (be it member variables or functions), should not be the first thing you see when you are looking at a class interface, anyway. Henceforth, I decided to keep this change as it is. It helps with readability: the user of Delta does not need to bother themselves with the knowledge that internally, there is some vector of state posts in `Delta`, which they cannot access or do anything with, anyway.

File: include/mata/nfa/delta.hh (hunk 2 above)
Comment (Adda0) [reply include/mata/nfa/delta.hh]: The `SeekFirst...` was replaced by a simple boolean flag parameter for constructor of `Moves`.

**Comment by Adda0** [comment]
>  I would rename const_iterator to ConstIterator to match the convention.

For renaming custom iterators, I have created a separate issue #312. I will keep the current naming convention in this PR. We may rename all custom iterators at once later on when resolving #312.

**Comment by Adda0** [comment]
All the issues and suggestions have been addressed. If you want to add anything more to this PR, let me know. Otherwise, when @kilohsakul reviews the code and his issues are resolved, the PR should be finished and ready to be merged.

File: include/mata/nfa/delta.hh (hunk 2 above)
Comment (Adda0) [reply include/mata/nfa/delta.hh]: Moved the logic getting begin and end symbol post iterators to `moves_...()` functions. Therefore, we cannot create a custom `Moves` class instance with a specified range of `[first_symbol, last_symbol]`. We can only create custom `Moves` class instance with iterators. If we want to specify the range with symbols, we need to use one of the helper functions for getting all labels, symbols, or epsilons.

**Comment by Adda0** [comment]
@kilohsakul What do you think about the latest changes, moving the code handling getting iterators to symbol posts to iterate over to functions `moves()`, `moves_epsilons()` and `moves_symbols()`? Is there anything more you would like to change, or do you approve the PR?

File: src/nfa/delta.cc [hunk 8]
```diff
@@ -340,82 +343,172 @@ void Delta::defragment(const BoolVector& is_staying, const std::vector<State>& r
...
+StatePost::Moves& StatePost::Moves::operator=(const Moves& other) noexcept {
+    if (&other != this) {
+        state_post_ = other.state_post_;
+        symbol_post_it_ = other.symbol_post_it_;
+        symbol_post_end_ = other.symbol_post_end_;
+    }
+    return *this;
+}
+
+StatePost::Moves StatePost::moves(
```
Comment (kilohsakul) [inline src/nfa/delta.cc:452]: Good, causes no brain demage now, nice.

**Review by kilohsakul — APPROVED** [review]

**Comment by Adda0** [comment]
PR has been rebased onto the current devel with refactored CMake. Once GAs finish, the PR will be merged.
